### DISCUSSION TITLE
Filament v3 Plans

### DISCUSSION THREAD

**Discussion by danharrin** [filamentphp/filament#3654] [General]
Filament v3 is on the horizon, and in the interest of transparency, I thought it would be a good idea to let you all know what our plans are.

## Livewire v3 only

According to the [Laracon website](https://laracon.net), Livewire v3 will be "introduced" on September 14th. I expect that it will be released during the next couple of months after that point. Since it is a new major version, it will bring breaking changes. Filament heavily relies on Livewire for all interactivity, so it is certain that we will experience these issues. It will not be easy to support two versions of Livewire at once, so we will need to introduce a new major version of Filament - v3. If we continued to only support Livewire v2 instead, anyone using Livewire v3 in their apps would be blocked from using Filament.

## Blade Heroicons v2 only

Blade Heroicons v2 was released today. Icons have been renamed / removed from the set, so we are unable to support both v1 and v2 of this package simultaneously. Therefore, Filament v3 will only support Blade Heroicons v2.

## Laravel 9 only

Blade Heroicons v2 has dropped support for Laravel 8. Since our only option is to support Heroicons v2 only, we must also drop support for Laravel 8. This was not our original plan, we were going to support Laravel 8 until Jan 24 2023, when security fixes will stop being issued.

## The future of Filament v2

I realise that these changes are going to make quite a few people unhappy, I am very sorry about that. Hopefully you can see that this situation is out of our control.

When we released v2, we immediately dropped support for new v1 features, but we did continue to release bug fixes and security fixes if a PR was submitted. This was a harsh approach, and I don't feel comfortable also doing this for our v3 upgrade.

Therefore, we will strive to continue **Filament v2 support with new features as well until at least Jan 24 2023, or when Livewire v2 stops receiving security fixes, whichever is earlier**. At that point, Laravel 8 or Livewire 2 apps will be vulnerable anyway. If it's easy for us to continue supporting that version past Jan 24 2023, we will do that. It doesn't look like much of Filament is going to change architecturally between v2 and v3, so that is good sign.

Going forward, we will aim to keep Filament major versions in line with Livewire's - Filament v4 and Livewire v4 may also be released simultaneously. If Livewire isn't planning regular major releases after v3, then we will abandon that plan. It makes sense for us to use Livewire as our marker for releases, as it is much more likely to introduce difficult breaking changes than Laravel is. Also, since a lot of Filament admin users also use the TALL stack, we don't want to hold them back by preventing them from installing the latest and greatest Livewire versions.

I'm going to keep this thread updated with new updates about the release, so subscribe to receive notifications!

**Comment by danharrin** [comment]
Last week, Caleb Porzio announced a ton of new Livewire v3 features at Laracon!

So, I thought it would be useful to go through a few of the features and let you know how they will improve your experience with Filament!

Jason Beggs has [written a post](https://laravel-news.com/livewire-v3-features) about the new features for Laravel News! I'll be quoting that throughout this post.

### @teleport

> Livewire v3 will also include a new @teleport Blade directive that will allow you to "teleport" a piece of markup and render it another part of the DOM. This can sometimes help avoid z-index issues with modals and slideouts.

In Filament v2, you are unable to embed the table builder within the form builder. The most useful application for this is putting a relation manager inside a resource form in the admin panel. This is due to a limitation with HTML, as it doesn't allow nested `<form>` elements.

This new teleport feature will allow us to teleport the nested `<form>` elements away, and provide you with this feature!

### Lazy components

> Livewire v3 will introduce "lazy" components. If you have a component that takes a while to load due to some expensive query or is rendered in a slideout that isn't always opened, you might want to wait to load it until it's shown on the page. In Livewire v3, just add a lazy attribute when rendering a component and it won't be rendered initially.

You may notice that on pages with lots of nested Livewire components (widgets and relation managers), the initial page load speed might take a hit. No more! This new feature will allow us to lazy-load in only the content after the initial request.

### Reactive properties

> When using nested components in Livewire v2, if a property on the parent component is updated, the child component's data is not automatically kept in sync. In Livewire v3, when you pass a piece of data to a child component, add a /** @prop */ comment above the property in the child, then update it in the parent component, it will be updated in the child component.

These new reactive properties will allow us to...

- Allow resource widgets to table filters! For instance, if you have an orders table with some filters, and an orders chart above it, the chart can be configured to only show the orders in the table when it's filtered!
- Provide Spatie Translatable content support for relation managers and widgets, since we can now pass the locale from the page to the nested Livewire components!

### Requests are batched

> In Livewire v2, if you have multiple components using wire:poll or dispatching and listening for events, each one of those components will send separate requests to the server on each poll or event. In Livewire v3, there is intelligent batching of requests so that wire:polls, events, listeners, and method calls can be batched into one request when possible, saving even more requests and improving performance.

This will be especially useful in the Filament dashboard, as each widget is its own Livewire component, and if you use polling, then the performance will be greatly improved.

### wire:navigate

> In Livewire v3, you'll be able to add wire:navigate to any anchor tag, and when clicked, Livewire will fetch the page in the background, then swap the DOM out really quickly giving your app a more SPA-like feel.

This will allow us to turn the admin panel into an SPA! Navigating between pages will feel faster than ever.

**Reply by tzsk** [reply]
All while looking dope AF. 🔥

**Reply by lloricode** [reply]
🔥🔥🔥🔥🔥🔥

**Reply by RizwanNaasir** [reply]
wire:navigate is gonna be game changer.  🤞

**Reply by mackhankins** [reply]
woah wire:navigate is fire

**Comment by danharrin** [comment]
Important proposal for v3, allowing you to build SaaS apps with `filament/filament`: https://github.com/filamentphp/filament/discussions/4400

**Comment by nnq1** [comment]
I hope this is an updated info "we will strive to continue Filament v2 support with new features as well until at least **Jan 24 2023**", as it tells the release is close. I would like to use Filament v3 ASAP.

**Reply by danharrin** [reply]
No thats not a release date or anything. Thats the date when Laravel 8 will stop receiving security fixes.

The reason why we would continue to add features Filament v2 while Filament v3 was released, is so that Laravel 8 users arent left in the dark. But if their apps are vulnerable anyway, theres not much point in making an effort. When I wrote this article, I thought we'd have released everything by now.

**Comment by ariaieboy** [comment]
You must mention the PHP version support too.
because of the `html-sanitizer` package filament must require PHP 8.1+ as the minimum requirement https://github.com/symfony/html-sanitizer/blob/6.1/composer.json#L19

**Comment by jeremiahsherrill** [comment]
I assume we are still waiting on livewire 3 to be released

**Reply by zepfietje** [reply]
Correct.

**Reply by fabianmuema** [reply]
Any idea on when it might be released?

**Reply by panakour** [reply]
> Any idea on when it might be released?

my prediction is that is going to take at least 2-3 months from now for livewire 3 to be released.

**Reply by zepfietje** [reply]
There's no ETA and we don't have any data on it. Let's not bloat this thread with guesswork.

**Reply by alvleont** [reply]
My prediction is that is going to be released as soon they consider it should be released. Or maybe a day after.

**Reply by Larry-Home** [reply]
I predict July 20th as announced :)

**Comment by donnyxray** [comment]
Do we have assurances that heroicons won't break backwards compatibility again in the future? Does it really make sense to keep using them?

I know some will argue that dependencies breaking compatibility is a fact of life. In my experience that's really quite rare though. And for an icon pack it just seems silly.

Keep in mind that many projects look at filament's choice of icons and re-use those in other areas to maintain a uniform appearance. The 3.x release will trigger a lot of extra work for something as basic as icons.

**Reply by danharrin** [reply]
The 3.x release has a script which is able to replace all old icons with their new names. Minimal work involved for people who are upgrading.

**Comment by alvleont** [comment]
https://v3when.com/


*Alvaro León Torres*
Químico Farmacéutico - UdeA
Telf: +57-305-263-3211
Integrante Grupo de PyP Farmacéutica - Universidad de Antioquia
Farmacéutico por pasión
Colombia.

**Comment by zircuitz** [comment]
Is there any way it would be possible to use custom icons? Why does it have to be tied to Heroicons? It would be so awesome if it was possible to use something like Font Awesome, or custom from SVG files.

**Reply by zepfietje** [reply]
v3 supports using custom icons or icon sets!

**Comment by alvleont** [comment]
You can use them. Just install the corresponding laravel extension.

**Reply by zircuitz** [reply]
You mean the [Blade plugin](https://github.com/owenvoke/blade-fontawesome) for Font Awesome? Excuse my expression, but it's total **trash**. The boot time slows to a crawl because it has to load everything after every request. Not to mention that it requires manual updates if you need to use a brand new icon that has not been added to the package yet.

**Reply by donnyxray** [reply]
AFAIK v3 uses the ```@svg``` helper, so you can disable the terribly slow blade-icon components in your blade-icons.php:

```php
return [
    'components' => [
        'disabled' => true,
    ],
];
```

**Reply by zepfietje** [reply]
v3 sets that config option for you in a middleware already. 👍

**Comment by alvleont** [comment]
You’re always invited to make your own. The GPL community will appreciate
it.

Btw you gave me an idea for a plugin but I’ll be for v3.

**Comment by Jamesking56** [comment]
How do we install v3 alpha right now?

**Reply by zepfietje** [reply]
Not. The beta will be released soon: https://v3when.com.

**Reply by nitinatvaluelabs** [reply]
Is a doc or anything available as of now? (I meant, maybe in the git repo?)

**Reply by mohamedsabil83** [reply]
You can move to [3.x branch](https://github.com/filamentphp/filament/tree/3.x/packages) and check either **docs folder** or **README** available inside each package folder. I believe it's not the final but you can know much
